pkg/fsp/header.go

```diff
@@ -0,0 +1,231 @@
+	s += fmt.Sprintf("Image ID                    : %s\n", ih.ImageID)
+	s += fmt.Sprintf("Image Size                  : %#08x %d\n", ih.ImageSize, ih.ImageSize)
+	s += fmt.Sprintf("Image Base                  : %#08x %d\n", ih.ImageBase, ih.ImageBase)
+	s += fmt.Sprintf("Image Attribute             : %s\n", ih.ImageAttribute)
+	s += fmt.Sprintf("Component Attribute         : %s\n", ih.ComponentAttribute)
+	s += fmt.Sprintf("Cfg Region Offset           : %#08x %d\n", ih.CfgRegionOffset, ih.CfgRegionOffset)
+	s += fmt.Sprintf("Cfg Region Size             : %#08x %d\n", ih.CfgRegionSize, ih.CfgRegionSize)
+	s += fmt.Sprintf("Reserved2                   : %#08x\n", ih.Reserved2)
+	s += fmt.Sprintf("TempRAMInit Entry Offset    : %#08x %d\n", ih.TempRAMInitEntryOffset, ih.TempRAMInitEntryOffset)
+	s += fmt.Sprintf("Reserved3                   : %#08x\n", ih.Reserved3)
+	s += fmt.Sprintf("NotifyPhase Entry Offset    : %#08x %d\n", ih.NotifyPhaseEntryOffset, ih.NotifyPhaseEntryOffset)
+	s += fmt.Sprintf("FSPMemoryInit Entry Offset  : %#08x %d\n", ih.FSPMemoryInitEntryOffset, ih.FSPMemoryInitEntryOffset)
+	s += fmt.Sprintf("TempRAMExit Entry Offset    : %#08x %d\n", ih.TempRAMExitEntryOffset, ih.TempRAMExitEntryOffset)
+	s += fmt.Sprintf("FSPSiliconInit Entry Offset : %#08x %d\n", ih.FSPSiliconInitEntryOffset, ih.FSPSiliconInitEntryOffset)
+	return s
+}
+
+// ImageRevision is the image revision fielf of the FSP info header.
+type ImageRevision uint32
+
+func (ir ImageRevision) String() string {
+	return fmt.Sprintf("%d.%d.%d.%d",
+		(ir>>24)&0xff,
+		(ir>>16)&0xff,
+		(ir>>8)&0xff,
+		ir&0xff,
+	)
+}
+
+// SpecVersion represents the spec version as a packed BCD two-digit,
+// dot-separated unsigned integer.
+type SpecVersion uint8
+
+func (sv SpecVersion) String() string {
+	return fmt.Sprintf("%d.%d", (sv>>4)&0x0f, sv&0x0f)
+}
+
+// ImageAttribute represents the image attributes.
+type ImageAttribute uint16
+
+func (ia ImageAttribute) String() string {
+	ret := fmt.Sprintf("%#04x GraphicsDisplaySupported", uint16(ia))
+	if uint16(ia) & ^(uint16(1)) != 0 {
+		ret += " (reserved bits are not zeroed)"
+	}
+	return ret
+}
+
+// ComponentAttribute represents the component attribute.
+type ComponentAttribute uint16
+
+func (ca ComponentAttribute) String() string {
+	var attrs []string
+	if uint16(ca)&0x01 == 1 {
+		attrs = append(attrs, "ReleaseBuild")
+	} else {
+		attrs = append(attrs, "DebugBuild")
+	}
+	if uint16(ca)&0x03 == 1 {
+		attrs = append(attrs, "OfficialRelease")
+	} else {
+		attrs = append(attrs, "TestRelease")
+	}
+	switch uint16(ca) >> 12 {
+	case 1:
+		attrs = append(attrs, "TypeFSP-T")
+	case 2:
+		attrs = append(attrs, "TypeFSP-M")
+	case 3:
+		attrs = append(attrs, "TypeFSP-S")
+	case 8:
+		attrs = append(attrs, "TypeFSP-O")
+	default:
+		attrs = append(attrs, "TypeFSPReserved")
+	}
+	ret := fmt.Sprintf("%#04x %s", uint16(ca), strings.Join(attrs, "|"))
+	// bits 2:12 are reserved
```

Comment:
Wait if bits 2:12 are reserved, how does case 8 above work?

Reply:
my bad, it's bits 11:2, the 8 above is correct

see https://www.intel.com/content/dam/www/public/us/en/documents/technical-specifications/fsp-architecture-spec-v2.pdf section 5.1.1 table 1